examples/game-of-life-challenge/src/game.rs

```diff
@@ -0,0 +1,1131 @@
+    /// the board and the number of steps, if we succeed on or before `max_steps`.
+    /// Otherwise return the last error.
+    pub fn advance_until(
+        &self,
+        conditions: &[Condition],
+        max_steps: u16,
+    ) -> Result<(u16, Self), (usize, ConditionFailureReason)> {
+        let mut board = self.clone();
+        let mut i = 0;
+        loop {
+            let result = board.check_conditions(conditions);
+            match result {
+                Ok(()) => {
+                    return Ok((i, board));
+                }
+                Err(error) => {
+                    if i == max_steps {
+                        return Err(error);
+                    }
+                    i += 1;
+                    board = board.advance_once();
+                }
+            }
+        }
+    }
+
+    fn index(&self) -> DirectBoard {
+        let mut index = BTreeMap::<_, BTreeSet<_>>::new();
+        for Position { x, y } in self.live_cells.iter().cloned() {
+            index.entry(x).or_default().insert(y);
+        }
+        DirectBoard {
+            size: self.size,
+            index,
+        }
+    }
+
+    /// Check that the board satisfies a given set of conditions.
+    pub fn check_conditions(
+        &self,
+        conditions: &[Condition],
+    ) -> Result<(), (usize, ConditionFailureReason)> {
+        self.index().check_conditions(conditions)
+    }
+
+    /// Check that the board satisfies the given puzzle.
+    pub fn check_puzzle(&self, puzzle: &Puzzle, steps: u16) -> Result<(), InvalidSolution> {
+        if self.size != puzzle.size {
+            return Err(InvalidSolution::SizeMismatch {
+                board_size: self.size,
+                puzzle_size: puzzle.size,
+            });
+        }
+        if steps < puzzle.minimal_steps || steps > puzzle.maximal_steps {
+            return Err(InvalidSolution::StepsOutOfRange {
+                steps,
+                min_steps: puzzle.minimal_steps,
+                max_steps: puzzle.maximal_steps,
+            });
+        }
+        if let Err((condition_index, reason)) = self.check_conditions(&puzzle.initial_conditions) {
+            return Err(InvalidSolution::InitialConditionFailed {
+                condition_index,
+                reason,
+            });
+        }
+        let final_board = self.advance(steps);
+        if let Err((condition_index, reason)) =
+            final_board.check_conditions(&puzzle.final_conditions)
+        {
+            Err(InvalidSolution::FinalConditionFailed {
+                condition_index,
+                steps,
+                reason,
+            })
+        } else {
+            Ok(())
+        }
+    }
+}
+
+impl Cell {
+    fn set_live(&mut self) {
+        self.value |= 1;
+    }
+
+    fn increment_neighbor_count(&mut self) {
+        self.value += 2;
+    }
+
+    fn should_be_live_next(self) -> bool {
+        self.value >= 5 && self.value <= 7
```

Comment:
No explaining comment? :sweat_smile: 
OK, I guess 5 means "live and 2 neighbors", 6 is "dead and 3 neighbors" and 7 is "live and 3 neighbors".